text/0550-macro-future-proofing.md

```diff
@@ -68,45 +92,329 @@ proposal is to prevent such scenarios in the future by requiring certain
+                           |                      |               |
+    FIRST:          { $d:ident }               { h, ε }         { f }
+
+    INPUT:  $(  $d:ident   $e:expr   );*    $( $( h )* );*    $( f ; )+   g
+            ~~~~~~~~~~~~~~~~~~~~~~~~~~~~    ~~~~~~~~~~~~~~    ~~~~~~~~~   ~
+                           |                       |              |       |
+    FIRST:        { $d:ident, ε }            {  h, ε, ;  }      { f }   { g }
+
+
+    INPUT:  $(  $d:ident   $e:expr   );*    $( $( h )* );*    $( f ; )+   g
+            ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
+                                            |
+    FIRST:                       { $d:ident, h, ;,  f }
+
+Thus:
+
+ * FIRST(`$($d:ident $e:expr );* $( $(h)* );* $( f ;)+ g`) = { `$d:ident`, `h`, `;`, `f` }
+
+Note however that:
+
+ * FIRST(`$($d:ident $e:expr );* $( $(h)* );* $($( f ;)+ g)*`) = { `$d:ident`, `h`, `;`, `f`, ε }
+
+Here are similar examples but now for LAST.
+
+ * LAST(`$d:ident $e:expr`) = { `$e:expr` }
+ * LAST(`$( $d:ident $e:expr );*`) = { `$e:expr`, ε }
+ * LAST(`$( $d:ident $e:expr );* $(h)*`) = { `$e:expr`, ε, `h` }
+ * LAST(`$( $d:ident $e:expr );* $(h)* $( f ;)+`) = { `;` }
+ * LAST(`$( $d:ident $e:expr );* $(h)* $( f ;)+ g`) = { `g` }
+ 
+ and again, changing the end part of matcher changes its last set considerably:
+ 
+ * LAST(`$( $d:ident $e:expr );* $(h)* $($( f ;)+ g)*`) = { `$e:expr`, ε, `h`, `g` }
+
+## FOLLOW(M)
+
+Finally, the definition for `FOLLOW(M)` is built up incrementally atop
+more primitive functions.
+
+We first assume a primitive mapping, `FOLLOW(NT)` (defined
+[below][follow-nt]) from a simple NT to the set of allowed tokens for
+the fragment specifier for that NT.
+
+Second, we generalize FOLLOW to tokens: FOLLOW(t) = FOLLOW(NT) if t is (a simple) NT.
+Otherwise, t must be some other (non NT) token; in this case FOLLOW(t) = ANYTOKEN.
+
+Finally, we generalize FOLLOW to arbitrary matchers by composing the primitive
+functions above:
+
+```
+FOLLOW(M) = FOLLOW(t1) ∩ FOLLOW(t2) ∩ ... ∩ FOLLOW(tN)
+            where { t1, t2, ..., tN } = (LAST(M) \ { ε })
+```
+
+Examples of FOLLOW (expressed as equality relations between sets, to avoid
+incorporating details of FOLLOW(NT) in these examples):
+
+ * FOLLOW(`$( $d:ident $e:expr )*`) = FOLLOW(`$e:expr`)
+ * FOLLOW(`$( $d:ident $e:expr )* $(;)*`) = FOLLOW(`$e:expr`) ∩ ANYTOKEN = FOLLOW(`$e:expr`)
+ * FOLLOW(`$( $d:ident $e:expr )* $(;)* $( f |)+`) = ANYTOKEN
+
+## FOLLOW(NT)
+[follow-nt]: #follownt
+
+Here is the definition for FOLLOW(NT), which maps every simple NT to
+the set of tokens that are allowed to follow it, based on the fragment
+specifier for the NT.
 
 The current legal fragment specifiers are: `item`, `block`, `stmt`, `pat`,
 `expr`, `ty`, `ident`, `path`, `meta`, and `tt`.
 
-- `FOLLOW(pat)` = `{FatArrow, Comma, Eq}`
+- `FOLLOW(pat)` = `{FatArrow, Comma, Eq, Or, Ident(if), Ident(in)}`
```

Comment:
(though I guess we would just not allow the generalized pattern form in that context?)